### PR TITLE
Toponaming: bring in getHigherElements code

### PR DESCRIPTION
**Pull request by bgbsww**
Seventh in series of incremental code transfers to solve the save/restore errors with objects with element maps..

### CODE REVIEW

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 1]
```diff
@@ -9603,6 +9695,69 @@ App::DocumentObject *SketchObject::getSubObject(
...
+            for (auto cstr : Constraints.getValues()) {
+                ++n;
+                if (cstr->Type != Sketcher::Coincident)
+                    continue;
+                for (int i=0; i<2; ++i) {
+                    int geoid = i ? cstr->Second : cstr->First;
+                    const Sketcher::PointPos &pos = i ? cstr->SecondPos : cstr->FirstPos;
+                    if(geoid >= 0 && index == getSolvedSketch().getPointId(geoid, pos) + 1)
+                        res.push_back(Data::IndexedName::fromConst("Constraint", n));
+                };
```
Comment (chennes) [inline src/Mod/Sketcher/App/SketchObject.cpp:9714]: I'm torn about this loop: on the one hand, DRY. On the other, this is kind of dumb complexity to achieve that. Maybe we should:
```suggestion
                if(cstr->First >= 0 && index == getSolvedSketch().getPointId(cstr->First, cstr->FirstPos) + 1)
                    res.push_back(Data::IndexedName::fromConst("Constraint", n));
                if(cstr->Second >= 0 && index == getSolvedSketch().getPointId(cstr->Second, cstr->SecondPos) + 1)
                    res.push_back(Data::IndexedName::fromConst("Constraint", n));
```
(On that note, does that `push_back` only ever get triggered once per constraint? Or can it happen that it hits for both `First` and `Second`?)

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 2]
```diff
@@ -9603,6 +9695,69 @@ App::DocumentObject *SketchObject::getSubObject(
...
+    if (res.empty()) {
+        res = { Part::TopoShape::shapeName(TopAbs_VERTEX).c_str(),
+                Part::TopoShape::shapeName(TopAbs_EDGE).c_str(),
+                "ExternalEdge",
+                "Constraint",
+                "InternalEdge",
+                "InternalFace",
+                "InternalVertex",
+              };
+    }
```
Comment (chennes) [inline src/Mod/Sketcher/App/SketchObject.cpp:9757]: No need to separate the logic like that:
```suggestion
    static std::vector<const char *> res { Part::TopoShape::shapeName(TopAbs_VERTEX).c_str(),
                Part::TopoShape::shapeName(TopAbs_EDGE).c_str(),
                "ExternalEdge",
                "Constraint",
                "InternalEdge",
                "InternalFace",
                "InternalVertex",
              };
```

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 3]
```diff
@@ -9603,6 +9695,69 @@ App::DocumentObject *SketchObject::getSubObject(
...
+                    int geoid = i ? cstr->Second : cstr->First;
+                    const Sketcher::PointPos &pos = i ? cstr->SecondPos : cstr->FirstPos;
+                    if(geoid >= 0 && index == getSolvedSketch().getPointId(geoid, pos) + 1)
+                        res.push_back(Data::IndexedName::fromConst("Constraint", n));
+                };
+            }
+        }
+        return res;
+
+    auto getNames = [&](const char *element) {
```
Comment (chennes) [inline src/Mod/Sketcher/App/SketchObject.cpp:9719]: The CI linter doesn't complain about it, but mine does --Clang-Tidy: Lambdas that implicitly capture 'this' should not specify a capture default"
```suggestion
    auto getNames = [this, &silent, &res](const char *element) {
```

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 4]
```diff
@@ -9603,6 +9695,69 @@ App::DocumentObject *SketchObject::getSubObject(
...
+        bool internal = boost::starts_with(element, internalPrefix());
+        const auto &shape = internal ? InternalShape.getShape() : Shape.getShape();
+        for (const auto &indexedName : shape.getHigherElements(element+(internal?internalPrefix().size():0), silent)) {
```
Comment (chennes) [inline src/Mod/Sketcher/App/SketchObject.cpp:9722]: As long as we're making fixes anyway...
```suggestion
        for (const auto &indexedName : shape.getHigherElements(element+(internal?internalPrefix().size() : 0), silent)) {
```

**Review by chennes — APPROVED** [review]
All minor stuff: if you're OK with it all I can just commit the suggestions during the merge.

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 1 above)
Comment (bgbsww) [reply src/Mod/Sketcher/App/SketchObject.cpp:9714]: I *think* that both will get hit for something like a coincident constraint.

**Pull request merged**